File: backend/app/schemas/student.py
from datetime import date,datetime
from uuid import UUID

from pydantic import BaseModel


class StudentProfileResponse(BaseModel):
    user_id: UUID
    profile_id: UUID

    full_name: str
    college_email: str
    personal_email: str
    phone_number: str
    date_of_birth: date
    gender: str
    exam_roll_number: str

    degree: str
    batch_year: int
    graduation_year: int

    department_profile_id: UUID


class StudentTestAssignmentResponse(BaseModel):
    assignment_id: UUID
    test_id: UUID

    title: str
    description: str | None

    mode: str
    duration_minutes: int | None

    status: str

    assigned_at: datetime
    started_at: datetime | None
    submitted_at: datetime | None

    released_at: datetime | None


class StudentTestsResponse(BaseModel):
    tests: list[StudentTestAssignmentResponse]


# ============================================================
# 4C — TEST ATTEMPT
# ============================================================

class StudentAttemptQuestionResponse(BaseModel):
    question_id: UUID
    sequence_number: int

    question: dict
    marks: float | None

    answer: dict | None
    is_answered: bool


class StudentTestAttemptResponse(BaseModel):
    assignment_id: UUID
    test_id: UUID

    title: str
    description: str | None

    mode: str
    duration_minutes: int | None

    status: str

    started_at: datetime | None
    submitted_at: datetime | None

    questions: list[StudentAttemptQuestionResponse]


class StudentAnswerRequest(BaseModel):
    answer: dict | None


class StudentAnswerResponse(BaseModel):
    message: str
    question_id: UUID
    answer: dict | None


class StudentSubmitResponse(BaseModel):
    message: str
    assignment_id: UUID
    status: str
    submitted_at: datetime

File: frontend/src/api/student.ts
import { apiRequest } from "./client";

export interface StudentProfile {
  user_id: string;
  profile_id: string;

  full_name: string;
  college_email: string;
  personal_email: string;
  phone_number: string;
  date_of_birth: string;
  gender: string;
  exam_roll_number: string;

  degree: string;
  batch_year: number;
  graduation_year: number;

  department_profile_id: string;
}

export function getStudentProfile() {
  return apiRequest<StudentProfile>(
    "/api/student/me",
  );
}


export type StudentAssignmentStatus =
  | "ASSIGNED"
  | "IN_PROGRESS"
  | "SUBMITTED"
  | "MISSED";


export interface StudentTestAssignment {
  assignment_id: string;
  test_id: string;

  title: string;
  description?: string | null;

  mode:
    | "AUTOMATIC"
    | "MANUAL"
    | "ADAPTIVE";

  duration_minutes?: number | null;

  status: StudentAssignmentStatus;

  assigned_at: string;
  started_at?: string | null;
  submitted_at?: string | null;

  released_at?: string | null;
}


export interface StudentTestsResponse {
  tests: StudentTestAssignment[];
}


export function getStudentTests() {
  return apiRequest<StudentTestsResponse>(
    "/api/student/tests",
  );
}

export interface StudentAttemptQuestion {
  question_id: string;
  sequence_number: number;

  question: Record<string, unknown>;
  marks?: number | null;

  answer?: Record<string, unknown> | null;
  is_answered: boolean;
}


export interface StudentTestAttempt {
  assignment_id: string;
  test_id: string;

  title: string;
  description?: string | null;

  mode:
    | "AUTOMATIC"
    | "MANUAL"
    | "ADAPTIVE";

  duration_minutes?: number | null;

  status: StudentAssignmentStatus;

  started_at?: string | null;
  submitted_at?: string | null;

  questions: StudentAttemptQuestion[];
}


export interface StudentAnswerResponse {
  message: string;
  question_id: string;
  answer?: Record<string, unknown> | null;
}


export interface StudentSubmitResponse {
  message: string;
  assignment_id: string;
  status: StudentAssignmentStatus;
  submitted_at: string;
}


export function startStudentTest(
  assignmentId: string,
) {
  return apiRequest<StudentTestAttempt>(
    `/api/student/assignments/${assignmentId}/start`,
    {
      method: "POST",
    },
  );
}


export function getStudentTestAttempt(
  assignmentId: string,
) {
  return apiRequest<StudentTestAttempt>(
    `/api/student/assignments/${assignmentId}`,
  );
}


export function saveStudentAnswer(
  assignmentId: string,
  questionId: string,
  answer: Record<string, unknown> | null,
) {
  return apiRequest<StudentAnswerResponse>(
    `/api/student/assignments/${assignmentId}/questions/${questionId}/answer`,
    {
      method: "POST",
      body: JSON.stringify({
        answer,
      }),
    },
  );
}


export function submitStudentTest(
  assignmentId: string,
) {
  return apiRequest<StudentSubmitResponse>(
    `/api/student/assignments/${assignmentId}/submit`,
    {
      method: "POST",
    },
  );
}